torchao/quantization/autoquant.py

```diff
@@ -479,32 +491,57 @@ def autoquant_prehook(module, args, kwargs):
     )
 
     # access actual model of torch.compile wrapper if needed
-    if isinstance(model, torch._dynamo.eval_frame.OptimizedModule):
+    is_compiled = isinstance(model, torch._dynamo.eval_frame.OptimizedModule)
+    if is_compiled:
         real_model = model._orig_mod
     else:
         real_model = model
 
-    # we need a consistent way to run the model which bypasses both
-    # A) the torch.compile tracing (so we need to run the inner model directly)
-    # B) the autoquant_prehook we're about to register (so we call forward directly)
-    model.forward_log_only = lambda *args, **kwargs: real_model.forward(*args, **kwargs)
-
-    # the autoquant_prehook intercepts the forward call and performs autoquantization
-    # and then deletes the hook. if model is a torch.compile wrapper, it then
-    # does the tracing/compile since the prehook is naturally followed by the normal.
-    # model run.
-    handle = model.register_forward_pre_hook(autoquant_prehook, with_kwargs=True)
+   
+    if manual_do_autoquant:
+        # we don't want model.forward to trigger
+        # torch.compilation
+        if is_compiled:
+            real_model.old_forward = model.forward
+            model.forward = real_model.forward
 
-    # note the torch.compile wrapper eval_frame moved the assignment of any assigned
-    # attributes to the inner model, so we have to call delattr on the inner model
-    def clean_up_autoquant_hooks_and_attrs():
-        try:
-            handle.remove()
-            delattr(real_model, "clean_up_autoquant_hooks_and_attrs")
+    # we want to automatically do autoquant after a single model run
+    # and have it occur before torch.compilation if applicable
+    else:
+        # the hook we will use to intercept the model forward and perform
+        # autoquantization
+        def autoquant_prehook(module, args, kwargs):
+            real_model.forward(*args, **kwargs)
+            module.do_autoquant()
+            return args, kwargs
+
+        # the autoquant_prehook intercepts the forward call, performs logging then
+        # does autoquantization. if model is a torch.compile wrapper, it then
+        # does the tracing/compile since the prehook is naturally followed by the normal.
+        # model run.
+        handle = model.register_forward_pre_hook(autoquant_prehook, with_kwargs=True)
+
+    # provide a method to bypass the prehook/torch.compilation
+    real_model.forward_log_only = real_model.forward
+        
+    # note the torch.compile wrapper (eval_frame) moves the assignment of any assigned
```

Comment:
one option is `model.compile()` instead of `torch.compile()` model which doesnt do any shenanigans with the module structure

Reply:
we still want to support torch.compile though which is the more common API